app/helpers/lti_helper.rb

```diff
@@ -0,0 +1,151 @@
+        { user_name: user['lis_person_sourcedid'].nil? ? user['name'] : user['lis_person_sourcedid'],
+          first_name: user['given_name'],
+          last_name: user['family_name'],
+          display_name: user['name'],
+          email: user['email'],
+          lti_user_id: user['user_id'] }
+      end
+    end
+    if user_data.empty?
+      raise I18n.t('lti.no_users')
+    end
+    user_data.each do |lms_user|
+      markus_user = EndUser.find_by(user_name: lms_user[:user_name])
+      if markus_user.nil? && can_create_users
+        markus_user = EndUser.create(lms_user.except(:lti_user_id))
+        if markus_user.nil?
+          error = true
+          next
+        end
+      elsif markus_user.nil? && !can_create_users
+        error = true
+        next
+      end
+      course_role = Student.find_by(user: markus_user, course: course)
+      if course_role.nil? && can_create_roles
+        course_role = Student.create!(user: markus_user, course: lti_deployment.course)
+      end
+      next if course_role.nil?
+      lti_user = LtiUser.find_or_initialize_by(user: markus_user, lti_client: lti_deployment.lti_client)
+      lti_user.update!(lti_user_id: lms_user[:lti_user_id])
+    end
+    error
+  end
+
+  def grade_sync(lti_deployment, assessment)
+    scopes = [LtiDeployment::LTI_SCOPES[:score], LtiDeployment::LTI_SCOPES[:results]]
+    create_or_update_lti_assessment(lti_deployment, assessment)
+    auth_data = lti_deployment.lti_client.get_oauth_token(scopes)
+    line_item = lti_deployment.lti_line_items.find_by!(assessment: assessment)
+    curr_results = get_current_results(line_item, auth_data, scopes)
+    score_uri = URI("#{line_item.lti_line_item_id}/scores")
+    req = Net::HTTP::Post.new(score_uri)
+
+    if assessment.is_a?(Assignment)
+      marks = get_assignment_marks(lti_deployment, assessment)
+    else
+      marks = get_grade_entry_form_marks(lti_deployment, assessment)
+    end
+    if marks.empty?
+      raise I18n.t('lti.no_grades')
+    end
+    marks.each do |lti_user_id, mark|
+      # Only send if the mark has not been previously sent to the LMS
+      # or if the mark differs from the LMS mark.
+      marked_by_lms = curr_results.find { |result| result['userId'] == lti_user_id }
+      if marked_by_lms.nil? || marked_by_lms['resultScore'] != mark
+        payload = {
+          timestamp: Time.current.iso8601,
+          scoreGiven: mark,
+          scoreMaximum: assessment.max_mark.to_f,
+          activityProgress: 'Completed',
+          gradingProgress: 'FullyGraded',
+          userId: lti_user_id
+        }
+        req.set_form_data(payload)
+        lti_deployment.send_lti_request!(req, score_uri, auth_data, scopes)
+      end
+    end
+  end
+
+  # Returns a hash mapping lti_user_id to marks
+  # for each released mark where the user has an lti_user_id
+  def get_assignment_marks(lti_deployment, assignment)
+    released_marks = assignment.released_marks
```

Comment:
This variable name is misleading, the result that's stored isn't mark data, it's result data.
In addition to renaming it, I recommend making this a hash mapping lti user id to result id.

Also (about the code below this line):

1. Whenever dealing with marks, we need to include only *accepted* memberships, never pending ones. `Grouping` has a `:accepted_student_memberships` association that you can join on directly.
2. You can put all the joins on one line. Syntax like `.joins(x: { y: { z: ... } } }` should work fine.
3. Rather than iterate over `grades`, I recommend iterating over the return hash of this query. Since `grades` is itself a hash, you won't need to call `.find`, but instead just use hash lookup to get the associated total grade.

Reply:
(Same general comments apply to `get_grade_entry_form_marks`